lib/internal/modules/esm/module_job.js

```diff
@@ -324,17 +324,17 @@ class ModuleJob extends ModuleJobBase {
     let status = this.module.getStatus();
 
     debug('ModuleJob.runSync()', status, this.module);
-    // FIXME(joyeecheung): this cannot fully handle < kInstantiated. Make the linking
-    // fully synchronous instead.
     if (status === kUninstantiated) {
-      this.module.hasAsyncGraph = this.module.instantiateSync();
+      // FIXME(joyeecheung): this cannot fully handle < kInstantiated. Make the linking
+      // fully synchronous instead.
+      assert(this.module.getModuleRequests().length === 0);
+      this.module.link([]);
```

Comment:
```suggestion
```

I think we can just drop this, even if there are module requests, it's not impossible that this job has already called `#link()` at this point and just needs an `instantiate()`. If it does, `instantiate()` would just work, and it's better than refusing to make it work. If it doesn't, `instantiate()` would throw `THROW_ERR_VM_MODULE_LINK_FAILURE` (maybe not the best error code considering it's not actually a VM module, but the error message is clearer and whether it's actually an VM module is more of an implementation detail).

Reply:
This `link([])` is required because the semantic of `instantiate()` requires the module to have `link()` called first.

> If it doesn't, instantiate() would throw THROW_ERR_VM_MODULE_LINK_FAILURE.

This is the exact case that the test case `test/es-module/test-import-require-tla-twice.js` is broken.